packages/teams-js/src/private/externalAppCommands.ts

```diff
@@ -0,0 +1,192 @@
+    items?: MessageExtensionItem[] | null;
+    /**
+     * text will exist only when type is message
+     */
+    text?: string | null;
+  }
+
+  /**
+   * @hidden
+   * The result type for the ActionCommandResponse
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export type ResultTypes = 'result' | 'message';
+
+  /**
+   * @hidden
+   * The payload of attachment item
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export interface MessageExtensionItem {
+    card: Card;
+    previewCard?: Card | null;
+  }
+
+  /**
+   * @hidden
+   * The payload of Card
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export interface Card {
+    appIcon?: string | null;
+    appId?: string | null;
+    appName?: string | null;
+    cardClientId?: string | null;
+    content: string;
+    contentType: string;
+    locale?: string | null;
+    name?: string | null;
+    userSpecificContent?: UserSpecificCardContent | null;
+    sourceType?: CardSourceType | null;
+  }
+
+  /**
+   * @hidden
+   * The payload that is used when executing an Adaptive Card Action.Submit
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export interface UserSpecificCardContent {
+    content?: string | null;
+    timeStamp: number;
+    version?: string | null;
+    authBlockLastUsedOn?: number | null;
+  }
+
+  /**
+   * @hidden
+   * The payload that is used for CardSourceType
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export type CardSourceType =
+    | 'ZeroInstallAppInstalledAppLinkUnfurl'
+    | 'NonZeroInstallAppInstalledAppLinkUnfurl'
+    | 'ZeroInstallAppNotInstalledAppLinkUnfurl'
+    | 'URLP'
+    | 'URLPRichInfo'
+    | 'NonLinkUnfurling';
+
+  /**
+   *
+   * @hidden
+   * Error that can be thrown from IExternalAppCommandsService.processActionCommand
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export interface ActionCommandError {
+    errorCode: string;
+    message?: string;
+  }
+
+  /**
+   * @internal
+   * Limited to Microsoft-internal use
+   * @hidden
+   * This API will delegate an ActionCommand request to the extensibility provider to send the request to the backend service of the application with provided:
+   *
+   * @param appId ID of the application the request is intended for. This must be a UUID
+   * @param commandId extensibilityProvider use this ID to look up the command declared by ActionME
+   * @param extractedParameters is the parameters that the dialog will be prepopulated with
+   * @param fromElement if `fromElemetn` is undefined, send the botME as a default value
+   * @returns Promise that resolves when the request is completed and rejects with ActionCommandError if the request fails
+   *
+   * @beta
+   */
+  export function processActionCommand(
+    appId: string,
+    commandId: string,
```

Comment:
>string [](http://example.com/codeflow?start=15&length=6)

Can this command id really be any string? Is "Maggie" a valid command id?

Reply:
Commands are defined by an external developer. We validate it in web-SDK by making sure the commandId from this API is defined in manifest. If the commandId comes in that is not defined in manifest (and hence in app definition), web-SDK throws an error. Also, we cannot valid the commandId here, because we don't have access to the external app definition in teams-js.